File: exercises/lru_cache_146.go
package main

type Node struct {
    val int
    next *Node
    prev *Node
}

type LinkedList struct {
    size int
    head *Node
    tail *Node
}

type LRUCache struct {
    capacity int
    list LinkedList
    lookup map[int]*Node
    reverseLookup map[*Node]int
}


func Constructor(capacity int) LRUCache {
    return LRUCache {capacity: capacity, list: LinkedList{size: 0}, 
        lookup: make(map[int]*Node), reverseLookup: make(map[*Node]int)}
}


func (this *LRUCache) Get(key int) int {
    node, ok := this.lookup[key]
    if !ok {
        return -1
    }
    this.MoveFront(node)
    return node.val
}

func (this *LRUCache) Put(key int, value int)  {
    node, ok := this.lookup[key]
    if ok {
        node.val = value
        this.MoveFront(node)
    } else if this.list.size == this.capacity {
        this.InsertFront(key, value)
        this.DeleteLast()
    } else {
        this.InsertFront(key, value)
    }
}

func (this *LRUCache) MoveFront(node *Node)  {
    if this.list.size < 2 {
        return
    }
    if this.list.head == node {
        return
    }
    if node.next != nil {
        node.next.prev = node.prev
    }
    if node.prev != nil {
        node.prev.next = node.next
    }
    if node == this.list.tail {
        this.list.tail = this.list.tail.prev
    }
    node.prev = nil
    node.next = this.list.head
    this.list.head.prev = node
    this.list.head = node
}

func (this *LRUCache) InsertFront(key int, value int) {
    node := &Node {val: value}
    node.next = this.list.head
    if this.list.head == nil {
        this.list.head = node
        this.list.tail = node
    } else {
        this.list.head.prev = node
        this.list.head = node
    }
    if this.list.tail != nil && this.list.tail.prev == nil {
        this.list.tail.prev = node
    }
    this.list.size++
    this.lookup[key] = node
    this.reverseLookup[node] = key
}

func (this *LRUCache) DeleteLast() bool {
    if this.list.size == 0 {
        return false
    }
    key := this.reverseLookup[this.list.tail]
    delete(this.lookup, key)
    delete(this.reverseLookup, this.list.tail)
    if this.list.tail != nil && this.list.tail.prev != nil {
        this.list.tail.prev.next = nil
        this.list.tail = this.list.tail.prev
    } else {
        this.list.head = nil
        this.list.tail = nil
    }
    this.list.size--
    return true
}

File: exercises/lru_cache_146.py
class Node:
    def __init__(self, value) -> None:
        self.value = value
        self.next = None
        self.prev = None

class LinkedList:
    def __init__(self, size) -> None:
        self.size = size
        self.head = None
        self.tail = None

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.list = LinkedList(0)
        self.lookup = {}
        self.reverseLookup = {}


    def get(self, key: int) -> int:
        if key not in self.lookup:
            return -1        
        node = self.lookup[key]
        self.moveFront(node)
        # for key, value in self.lookup.items():
        #     print(f"{key}: {value.value}", end=' <> ')
        # print("\n")
        return node.value
        

    def put(self, key: int, value: int) -> None:
        if key in self.lookup:
            node = self.lookup[key]
            node.value = value
            self.moveFront(node)
        elif self.list.size == self.capacity:
            self.insertFront(key, value)
            self.deleteLast()
        else:
            self.insertFront(key, value)
        # for key, value in self.lookup.items():
        #     print(f"{key}: {value.value}", end=' <> ')
        # print("\n")

    def moveFront(self, node):
        if self.list.size < 2:
            return
        if self.list.head == node:
            return
        if node.next is not None:
            node.next.prev = node.prev
        if node.prev is not None:
            node.prev.next = node.next

        if node is self.list.tail:
            self.list.tail = self.list.tail.prev
        node.prev = None
        node.next = self.list.head
        self.list.head.prev = node
        self.list.head = node
        

    def insertFront (self, key, value):
        new_node = Node(value)
        new_node.next = self.list.head
        if self.list.head is None:
            self.list.head = new_node
            self.list.tail = new_node
        else:
            self.list.head.prev = new_node
            self.list.head = new_node
        if self.list.tail is not None and self.list.tail.prev is None:
            self.list.tail.prev = new_node
        self.list.size += 1
        self.lookup[key] = new_node
        self.reverseLookup[new_node] = key

    def deleteLast(self) -> bool:
        if self.list.size == 0:
            return False

        key = self.reverseLookup[self.list.tail]
        del self.lookup[key]
        del self.reverseLookup[self.list.tail]
        if self.list.tail is not None and self.list.tail.prev is not None:
            self.list.tail.prev.next = None
            self.list.tail = self.list.tail.prev
        else:
            self.list.head = None
            self.list.tail = None
        self.list.size -= 1
        
        return True

